clippy_utils/src/format.rs

```diff
@@ -0,0 +1,49 @@
+use crate::last_path_segment;
+use if_chain::if_chain;
+use rustc_hir::QPath;
+use rustc_hir::{BorrowKind, Expr, ExprKind};
+use rustc_span::sym;
+
+pub fn is_display_arg(expr: &Expr<'_>) -> bool {
+    if_chain! {
+        if let ExprKind::Call(_, [_, fmt]) = expr.kind;
+        if let ExprKind::Path(QPath::Resolved(_, path)) = fmt.kind;
+        if let [.., t, _] = path.segments;
+        if t.ident.name.as_str() == "Display";
+        then { true } else { false }
+    }
+}
+
+/// Checks if the expression matches
+/// ```rust,ignore
+/// &[_ {
+///    format: _ {
+///         width: _::Implied,
+///         precision: _::Implied,
+///         ...
+///    },
+///    ...,
+/// }]
+/// ```
+pub fn check_unformatted(expr: &Expr<'_>) -> bool {
+    if_chain! {
+        if let ExprKind::AddrOf(BorrowKind::Ref, _, expr) = expr.kind;
+        if let ExprKind::Array([expr]) = expr.kind;
+        // struct `core::fmt::rt::v1::Argument`
+        if let ExprKind::Struct(_, fields, _) = expr.kind;
+        if let Some(format_field) = fields.iter().find(|f| f.ident.name == sym::format);
+        // struct `core::fmt::rt::v1::FormatSpec`
+        if let ExprKind::Struct(_, fields, _) = format_field.expr.kind;
+        if let Some(precision_field) = fields.iter().find(|f| f.ident.name == sym::precision);
+        if let ExprKind::Path(ref precision_path) = precision_field.expr.kind;
+        if last_path_segment(precision_path).ident.name == sym::Implied;
+        if let Some(width_field) = fields.iter().find(|f| f.ident.name == sym::width);
+        if let ExprKind::Path(ref width_qpath) = width_field.expr.kind;
+        if last_path_segment(width_qpath).ident.name == sym::Implied;
+        then {
+            return true;
+        }
+    }
+
+    false
+}
```

Comment:
Since this is moving to utils, there are some things to fix.

1. Make this a method of `FormatArgsExpn` and give it a "positive" name like `format_args.has_formatting()`
1. Update the docs to say something like "returns true if any formating parameters are used like `{:+2}` instead of just `{}`"
1. Work for any number of args - the `ExprKind::Array` at the top currently only works for one arg.
1. Check for `flags: 0` in addition to checking `precision` and `width`
    * For perf - avoid doing `fields.iter().find()` for every field

Reply:
> 4\. Check for `flags: 0` in addition to checking `precision` and `width`

`useless_format` doesn't seem to like it when we check for `flags: 0`. So, I'm proposing that `has_formatting` be told what formatting types to check for, e.g.:
```rust
bitflags! {
    pub struct Formatting: u32 {
        const FILL = 0b00000001;
        const ALIGN = 0b00000010;
        const FLAGS = 0b00000100;
        const PRECISION = 0b00001000;
        const WIDTH = 0b00010000;
    }
}
```
```rust
                if !format_args.has_formatting(Formatting::PRECISION | Formatting::WIDTH);
```
Is it ok to add `bitflags` as a dependency of `clippy_utils`?

Reply:
> `useless_format` doesn't seem to like it when we check for `flags: 0`.

It looks like `useless_format` does not consider `flags` because it happens that none of the "flags" have any effect when the value is a string, and `useless_format` only triggers on strings. What if we just have a `has_string_formatting` which returns true if there is formatting that has an effect for string values? The bitflags seems more complex than necessary, and their general usefulness is limited since they don't let you see actual values like width and alignment.

Reply:
I think I implemented what you described, but please let me know if I misunderstood.